File: starterAIs/Starter.pl
use strict;
use warnings;
use 5.32.1;
$|=1;

use constant ME => 1;
use constant OPP => 0;
use constant NONE => -1;

chomp(my $tokens=<>);
my ($width, $height) = split / /, $tokens;

my $turn = -1;
while (1) {
    $turn++;
    
    chomp($tokens=<>);
    my ($my_matter, $opp_matter) = split / /, $tokens;
    
    my (@tiles, @myTiles, @oppTiles, @neutralTiles, @myUnits, @oppUnits, @myRecyclers, @oppRecyclers);
    for my $j (0..$height-1) {
        for my $i (0..$width-1) {
            chomp($tokens=<>);
            my ($scrap, $owner, $units, $recycler, $can_build, $can_spawn, $in_range) = split / /, $tokens;
            my %tile = (
                x => $i,
                y => $j,
                scrap => $scrap,
                owner => $owner,
                units => $units,
                recycler => $recycler,
                canBuild => $can_build,
                canSpawn => $can_spawn,
                inRange => $in_range
            );
            
            push @tiles, \%tile;
            
            if ($tile{owner} == ME) {
                push @myTiles, \%tile;
                push @myUnits, \%tile if $tile{units} > 0;
                push @myRecyclers, \%tile if $tile{recycler};
            } elsif ($tile{owner} == OPP) {
                push @oppTiles, \%tile;
                push @oppUnits, \%tile if $tile{units} > 0;
                push @oppRecyclers, \%tile if $tile{recycler};
            } else {
                push @neutralTiles, \%tile;
            }
        }
    }
    
    my @actions;
    
    for my $tile (@myTiles) {
        if ($tile->{canSpawn}) {
            my $amount = 0;  #TODO: Pick amount of robots to spawn here.
            push @actions, "SPAWN $amount " . $tile->{x} . " " . $tile->{y} if $amount;
        }
        
        if ($tile->{canBuild}) {
            my $shouldBuild = 0;  #TODO: Pick whether to build recycler here.
            push @actions, "BUILD " . $tile->{x} . " " . $tile->{y} if $shouldBuild;
        }
    }
    
    for my $tile (@myUnits) {
        my $target = "";  #TODO: Pick a destination.
        if ($target) {
            my $amount = 0;  #TODO: Pick amount of units to move.
            push @actions, "MOVE $amount " . $tile->{x} . " " . $tile->{y} . " " . $target->{x} . " " . $target->{y};
        }
    }
    
   say @actions-0 == 0 ? "WAIT" : join ";", @actions;
}

File: starterAIs/Starter.rb
STDOUT.sync = true # DO NOT REMOVE

ME = 1
OPP = 0
NONE = -1

width, height = gets.split.map &:to_i

# game loop
loop {
  tiles = []
  my_units = []
  opp_units = []
  my_recyclers = []
  opp_recyclers = []
  opp_tiles = []
  my_tiles = []
  neutral_tiles = []

  my_matter, opp_matter = gets.split.map &:to_i
  height.times { |y|
    width.times { |x|
     # owner: 1 = me, 0 = foe, -1 = neutral
     scrap_amount, owner, units, recycler, can_build, can_spawn, in_range_of_recycler = gets.split.map &:to_i

     tile = {
       scrap_amount: scrap_amount,
       owner: owner,
       units: units,
       recycler: recycler,
       can_build: can_build==1,
       can_spawn: can_spawn==1,
       in_range_of_recycler: in_range_of_recycler==1
     }

     tiles.append(tile)

     if tile.owner == ME
         my_tiles.append(tile)
         if tile.units > 0
             my_units.append(tile)
         elsif tile.recycler
             my_recyclers.append(tile)
         end
     elsif tile.owner == OPP
         opp_tiles.append(tile)
         if tile.units > 0
             opp_units.append(tile)
         elsif tile.recycler
             opp_recyclers.append(tile)
         end
     else
         neutral_tiles.append(tile)
     end
    }
  }

  actions = []
  my_tiles.each { |tile|
    if tile.can_spawn
      amount = 0 # TODO: pick amount of robots to spawn here
      if amount > 0
          actions<<"SPAWN #{amount} #{tile.x} #{tile.y}"
      end
    end
    if tile.can_build
        should_build = false # TODO: pick whether to build recycler here
        if should_build
            actions<<"BUILD #{tile.x} #{tile.y}"
        end
    end
  }

  my_units.each { |tile|
    target = nil # TODO: pick a destination tile
    if taget
      amount = 0 # TODO: pick amount of units to move
      actions<<"MOVE #{amount} #{tile.x} #{tile.y} #{target.x} #{target.y}"
    end
  }

  # To debug: STDERR.puts "Debug messages..."
  puts actions.size > 0 ? actions*";" : "WAIT"
}